src/rewriter/rw_parser.py

```diff
@@ -18,128 +18,52 @@
-    '+': Op.ADD,
-    '-': Op.SUB,
-    '*': Op.MULT,
-    'div': Op.INT_DIV,
-    '/': Op.DIV,
-    'mod': Op.MOD,
-    'abs': Op.ABS,
-    '<': Op.LT,
-    '>': Op.GT,
-    '<=': Op.LEQ,
-    '>=': Op.GEQ,
-    '=': Op.EQ,
-    'ite': Op.ITE,
-    'str.++': Op.STRING_CONCAT,
-    'str.in_re': Op.STRING_IN_REGEXP,
-    'str.len': Op.STRING_LENGTH,
-    'str.substr': Op.STRING_SUBSTR,
-    'str.update': Op.STRING_UPDATE,
-    'str.at': Op.STRING_AT,
-    'str.contains': Op.STRING_CONTAINS,
-    'str.<': Op.STRING_LT,
-    'str.<=': Op.STRING_LEQ,
-    'str.indexof': Op.STRING_INDEXOF,
-    'str.indexof_re': Op.STRING_INDEXOF_RE,
-    'str.replace': Op.STRING_REPLACE,
-    'str.replace_all': Op.STRING_REPLACE_ALL,
-    'str.replace_re': Op.STRING_REPLACE_RE,
-    'str.replace_re_all': Op.STRING_REPLACE_RE_ALL,
-    'str.prefixof': Op.STRING_PREFIX,
-    'str.suffixof': Op.STRING_SUFFIX,
-    'str.is_digit': Op.STRING_IS_DIGIT,
-    'str.from_int': Op.STRING_ITOS,
-    'str.to_int': Op.STRING_STOI,
-    'str.to_code': Op.STRING_TO_CODE,
-    'str.from_code': Op.STRING_FROM_CODE,
-    'str.tolower': Op.STRING_TOLOWER,
-    'str.toupper': Op.STRING_TOUPPER,
-    'str.rev': Op.STRING_REV,
-    'str.to_re': Op.STRING_TO_REGEXP,
-    're.++': Op.REGEXP_CONCAT,
-    're.union': Op.REGEXP_UNION,
-    're.inter': Op.REGEXP_INTER,
-    're.diff': Op.REGEXP_DIFF,
-    're.*': Op.REGEXP_STAR,
-    're.+': Op.REGEXP_PLUS,
-    're.opt': Op.REGEXP_OPT,
-    're.range': Op.REGEXP_RANGE,
-    're.comp': Op.REGEXP_COMPLEMENT,
-}
+symbol_to_op = {e.symbol: e for e in Op if e.symbol is not None}
 
 
 class SymbolTable:
     def __init__(self):
         self.consts = {
             're.none': App(Op.REGEXP_NONE, []),
+            're.all': App(Op.REGEXP_ALL, []),
             're.allchar': App(Op.REGEXP_ALLCHAR, []),
             '_': Placeholder(),
         }
         self.symbols = {}
+        # Definitions must be decoupled from symbols because symbols have to be
+        # bound as to not trigger an error from mkrewrites when it checks rule
+        # validity.
+        self.defs = {}
 
     def add_symbol(self, name, sort):
         if name in self.consts or name in self.symbols:
             die(f'Symbol {name} has already been declared')
 
         self.symbols[name] = Var(fresh_name(name), sort)
+    def add_def(self, name, expr):
+        if name in self.consts or name in self.symbols or name in self.defs:
+            die(f'Definition {name} has already been declared')
+
+        self.defs[name] = expr
 
     def get_symbol(self, name):
         if name in self.consts:
             return self.consts[name]
 
-        if name not in self.symbols:
+        if name not in self.symbols and name not in self.defs:
+            raise RuntimeError(f'Symbol {name} not declared')
             die(f'Symbol {name} not declared')
-        return self.symbols[name]
+        # Should not matter the order of defs and symbols since collisions are
+        # detected.
```

Comment:
Not sure what this comment refers to.

Reply:
Ok, then clarify the comment what this means. Essentially this means that get_symbol() will return the correct symbol that is currently in scope?